Mathlib/Algebra/Lie/LieTheorem.lean

```diff
@@ -0,0 +1,554 @@
+  Submodule.span_mono (fun _ ⟨c, hc, hw⟩ ↦ ⟨c, lt_of_lt_of_le hc h, hw⟩)
+
+lemma map_U_le (n : ℕ) : Submodule.map f (U v f n) ≤ U v f (n + 1) := by
+  rw [Submodule.map_span]
+  apply Submodule.span_mono
+  intro w ⟨z, ⟨m, hm, hz⟩, hw⟩
+  use m + 1, Nat.add_lt_add_right hm 1
+  rw [← hw, hz, ← LinearMap.mul_apply, ← pow_succ']
+
+def U_map : ℕ →o Submodule F V where
+  toFun := U v f
+  monotone' _ _ := U_mono v f
+
+abbrev iSupU : Submodule F V := ⨆ k : ℕ, U_map v f k
+
+lemma iSupU_eq_U : ∃ N : ℕ, iSupU v f = U v f N := by
+  apply Submodule.FG.stabilizes_of_iSup_eq (IsNoetherian.noetherian (iSupU v f)) (U_map v f)
+  rfl
+
+lemma map_iSupU_le_iSupU: Submodule.map f (iSupU v f) ≤ iSupU v f := by
+  rcases iSupU_eq_U v f with ⟨N, hN⟩
+  nth_rewrite 1 [hN]
+  apply le_trans' $ le_iSup (U v f) (N + 1)
+  exact map_U_le v f N
+
+end
+
+section
+variable (k V)
+-- the lie action of `L` on `V`
+abbrev π (z : L) : Module.End k V := LieModule.toEnd k L V z
+variable {k V}
+
+variable (A : LieIdeal k L) (χ : Module.Dual k A)
+
+abbrev T (w : A) : Module.End k V := (π k V w)  - χ w • 1
+
+-- We define a submodule of V which is L-invariant
+def altWeightSpace : Submodule k V where
+  carrier := {v |∀ a : A, ⁅a.val,v⁆ = (χ a)•v}
+  add_mem' := by
+    intro _ _ hb hc _
+    rw [lie_add, hc, hb, ← smul_add]
+
+  zero_mem' := by
+    intro a
+    rw [lie_zero, smul_zero]
+  smul_mem' := by
+    intro _ _ hx _
+    rw [lie_smul, hx, smul_comm]
+
+variable (z : L) (w : A) {v : V}
+
+lemma T_apply_succ (hv : v ∈ altWeightSpace A χ) (n : ℕ) :
+    Submodule.map (T A χ w) (U v (π k V z) (n + 1)) ≤ U v (π k V z) n := by
+  rw [Submodule.map_span, Submodule.span_le]
+  revert w
+  induction n
+  · intro w x ⟨v', ⟨m, hm, hv'⟩, hx⟩
+    rw [Nat.lt_one_iff.mp hm, pow_zero, LinearMap.one_apply] at hv'
+    rw [hv', T, LinearMap.sub_apply, LieModule.toEnd_apply_apply,
+      LieIdeal.coe_bracket_of_module, LinearMap.smul_apply,
+      LinearMap.one_apply, hv w, sub_self] at hx
+    rw [← hx]; exact Submodule.zero_mem _
+  · next n hn =>
+    intro w x ⟨v', ⟨m, hm, hv'⟩, hx⟩
+    rcases (eq_or_lt_of_le (Nat.le_of_lt_succ hm)) with (rfl | hm')
+    · rw [← hx, hv', T, LinearMap.sub_apply]
+      rw [pow_succ', LinearMap.mul_apply]
+      rw [LieModule.toEnd_apply_apply, LieModule.toEnd_apply_apply,
+        LieIdeal.coe_bracket_of_module, LinearMap.smul_apply, LinearMap.one_apply,
+        SetLike.mem_coe, leibniz_lie, add_sub_assoc]
+      apply Submodule.add_mem
+      · let wz : A := ⟨⁅w, z⁆, lie_mem_left k L A w.val z w.prop⟩
+        have : ⁅⁅w.val, z⁆, (π k V z ^ n) v⁆ =
+            (T A χ wz) ((π k V z ^ n) v) + χ wz  • ((π k V z ^ n) v) := by
+          simp
+        rw [this]
+        apply Submodule.add_mem
+        · exact U_mono v _ (Nat.le_succ n)
+            (hn wz ⟨(π k V z ^ n) v, ⟨n, Nat.lt_succ_self n, rfl⟩, rfl⟩)
+        · exact Submodule.smul_mem _ _ (Submodule.subset_span ⟨n, Nat.lt_succ_self n, rfl⟩)
+      · have  : ⁅w.val, (π k V z ^ n) v⁆ =
```

Comment:
Does this fit on one line, as in ` have  : ⁅w.val, (π k V z ^ n) v⁆ = (T A χ w) ((π k V z ^ n) v) + χ w  • ((π k V z ^ n) v) := by`? (You also have some double spaces in this line; please remove those.)